proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,355 @@
+val item : List<Item> by _items
+```
+
+And the new syntax allows us to write:
+
+```kotlin
+val items: List<Item>
+    internal field = mutableListOf()
+```
+
+This use-case is also widely applicable to architecture of reactive applications:
+ 
+* Android `LiveData` has a `MutableLiveData` counterpart.
+* Rx `Observable` has a mutable `Subject` counterpart.
+* Kotlin coroutines `SharedFlow` has a `MutableSharedFlow`, etc.
+
+For example, sample code [from an Android app](https://github.com/elpassion/crweather/blob/9c3e3cb803b7e4fffbb010ff085ac56645c9774d/app/src/main/java/com/elpassion/crweather/MainModel.kt#L14):
+
+```kotlin
+private val mutableCity = MutableLiveData<String>().apply { value = "" }
+private val mutableCharts = MutableLiveData<List<Chart>>().apply { value = emptyList() }
+private val mutableLoading = MutableLiveData<Boolean>().apply { value = false }
+private val mutableMessage = MutableLiveData<String>()
+
+val city: LiveData<String> = mutableCity
+val charts: LiveData<List<Chart>> = mutableCharts
+val loading: LiveData<Boolean> = mutableLoading
+val message: LiveData<String> = mutableMessage
+```
+
+Becomes:
+
+```kotlin
+val city: LiveData<String>
+    field = MutableLiveData().apply { value = "" }
+val charts: LiveData<List<Chart>>
+    field = MutableLiveData().apply { value = emptyList() }
+val loading: LiveData<Boolean>
+    field = MutableLiveData().apply { value = false }
+val message: LiveData<String>
+    field = MutableLiveData()
+```
+        
+> `private` is a default access for a `field` declaration.
+
+In this use-case, an read access to the field from inside the corresponding classes/modules 
+(where the field is visible) automatically gives access to mutable type, which we call [Smart Type Narrowing](#smart-type-narrowing), 
+but is seen as a property with read-only type to the outside code.
+
+### Decouple storage type from external representation
+
+Sometimes a property must be internally represented by a different type for storage-efficiency or architectural reason,
+while having a different outside type. For example, API requirements might dictate that the property type
+is `String`, but if we know that it always represents a decimal integer, then it can be efficiently stored as such 
+with custom getter and custom setter.
+
+```kotlin
+val number: String
+    field: Int = 0
+    get() = field.toString()
+    set(value) { field = value.toInt() }
+```
+
+### Expose read-only view
+
+In some application it is desirable to expose not just a read-only subtype, but a specially constructed read-only
+view that protects the data structure from casting into mutable type. For example a `MutableStateFlow` has
+[`asStateFlow`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-state-flow.html)
+extension for that purpose.
+
+With the proposed syntax it is possible to declare a custom getter:
+
+```kotlin
+val state: StateFlow<State> 
+    get() = field.asStateFlow()
```

Comment:
This creates a view on each getter invocation which may be unwanted.

Reply:
Added clarification on this point.